hearth/lib/hearth/middleware/retry.rb

```diff
@@ -2,114 +2,93 @@
-      # @param [String] max_attempts An integer representing the maximum number
-      #   of attempts that will be made for a single request, including the
-      #   initial attempt.
-      # @param [Boolean] adaptive_retry_wait_to_fill Used only in `adaptive`
-      #   retry mode. When true, the request will sleep until there is
-      #   sufficient client side capacity to retry the request. When false,
-      #   the request will raise a `CapacityNotAvailableError` and will not
-      #   retry instead of sleeping.
-      def initialize(app, retry_mode:, max_attempts:,
-                     adaptive_retry_wait_to_fill:, error_inspector_class:,
-                     retry_quota:, client_rate_limiter:)
+      # @param [RetryStrategy] retry_strategy The retry strategy to use.
+      #   Hearth has two built in classes, Standard and Adaptive RetryStrategy.
+      #   * `StandardRetryStrategy` - A standardized set of retry rules across
+      #     the AWS SDKs. This includes support for retry quotas, which limit
+      #     the number of unsuccessful retries a client can make.
+      #   * `AdaptiveRetryStrategy` - An experimental retry mode that includes
+      #     all the functionality of `standard` mode along with automatic
+      #     client side throttling. This is a provisional mode that may change
+      #     behavior in the future.
+      def initialize(app, retry_strategy:, error_inspector_class:)
         @app = app
 
-        # public config
-        @retry_mode = retry_mode
-        @max_attempts = max_attempts
-        @adaptive_retry_wait_to_fill = adaptive_retry_wait_to_fill
-
-        # undocumented options
+        @retry_strategy = retry_strategy
+        # undocumented - protocol specific
         @error_inspector_class = error_inspector_class
-        @retry_quota = retry_quota
-        @client_rate_limiter = client_rate_limiter
-
-        # instance state
-        @capacity_amount = nil
-        @retries = 0
       end
 
-      # @param input
-      # @param context
-      # @return [Output]
-      # rubocop:disable Metrics/AbcSize
       def call(input, context)
-        acquire_token
-        output = @app.call(input, context)
-        error_inspector = @error_inspector_class.new(
-          output.error, context.response.status
-        )
-        request_bookkeeping(output, error_inspector)
-
-        return output unless retryable?(context, output, error_inspector)
-
-        return output if @retries >= @max_attempts - 1
-
-        @capacity_amount = @retry_quota.checkout_capacity(error_inspector)
-        return output unless @capacity_amount.positive?
-
-        delay = [Kernel.rand * (2**@retries), MAX_BACKOFF].min
-        Kernel.sleep(delay)
-        retry_request(input, context, output)
+        token = @retry_strategy.acquire_initial_retry_token
+
+        raise 'Unable to make request, too many failures' if token.nil?
+
+        # TODO: determine if loop or nested function call is better here
```

Comment:
We used to invoke `call` over and over. Now it's a loop. A loop is less overhead. I'm open to either option.